src/qibo/gates/gates.py

```diff
@@ -2,11 +2,102 @@
 from typing import List
 
 import numpy as np
+import sympy as sp
 
 from qibo.config import PRECISION_TOL, raise_error
 from qibo.gates.abstract import Gate, ParametrizedGate
 
 
+class Parameter:
+    """Object which allows for variational gate parameters. Several trainable parameter
+    and possibly features are linked through a lambda function which returns the
+    final gate parameter
+
+    Args:
+        func (function): lambda function describing the gate parameter
```

Comment:
I think we should define a conventional way to write the lambdas. E.g., in your examples you always write functions like: 

```pyhton
lambda x1, x2, th1, th2 : x1 * th2 - x2 * th1
```

And this is fine with me, but I think it should be clear what is a feature and what is a parameter. For example, we can specify that the lambdas have to be provided writing features (x's) followed by parameters (th's). This can also help when calling the `update_parameters` function. 

I suggest to write these instructions in the docstrings:

```
    func (callable): lambda function which builds the gate parameter. If both features and trainable parameters 
        compose the function, it must be passed by first providing the features and then the parameters, as
        described in the code example above. 
```